### ISSUE TITLE
registerOnRemoveHandle is not allowed in the settings context

### ISSUE THREAD

**Issue by quesurifn** [OfficeDev/microsoft-teams-library-js#435]
I'm getting an error within teams like `registerOnRemoveHandle is not allowed in the settings context` when building an Office365 Connector

No matter where I place the handler, this happens. How does this get fixed?

Code: 

```
import React from 'react';
import TeamsBaseComponent from "msteams-react-base-component";
import CompletionImage from './images/completed-image.png';
import * as microsoftTeams from "@microsoft/teams-js";
import { Provider } from "@fluentui/react-northstar";
import { getAuthUrl } from './utils/api';
import OauthPopup from 'react-oauth-popup';
import * as service from './services/api';
import localStorage from 'local-storage';
import slugify from 'slugify';

class App extends TeamsBaseComponent {
  constructor() {
    super()
    this.state = {
      insideTeams: null,
      channelName: null,
      channelId: null,
      teamName: null,
      teamId: null,
      webhookUrl: null,
      authCode: null,
      flowStep: 'pre-authentication',
      error: false,
      isAlreadyAuthed: false,
      needsGraphAuth: false,
    }

    this.onCode = this.onCode.bind(this)
  }

  onClose() {
    if(process.env.NODE_ENV === 'development') {
      console.error("Modal Closed")
    }
  }

  debug(msg) {
    if(process.env.NODE_ENV === 'development') {
      console.debug(msg)
    }
  }

  async onCode(authCode, _params) {
    this.debug('On Code')
    const insideTeams = await this.inTeams()
    try {
      const {data} = await service.getAuthCode()

      if(insideTeams) {
        const responseObject = await service.finishChannelSetup(data.authentication_code, this.state)
        if (responseObject.success) {
          this.setState({channel: responseObject.channel, flowStep: 'authenticated',})
        } else {
          this.setState({needsGraphAuth: true, flowStep: false})
        }

      } else {
        this.setState({flowStep: 'authenticated', channel: true})
      }

    } catch(e) {
      this.debug(e)
      this.setState({error: e.message, flowStep: false})
    }
  }

  async componentWillMount() {
    const { error } = this.state
    const insideTeams = await this.inTeams()
    if(insideTeams) {
      microsoftTeams.initialize()
      microsoftTeams.getContext((context) => {
        this.setState({
            teamName: context.teamName,
            channelName: context.channelName,
            channelId: context.channelId,
            teamId: context.teamId,
            channelIsAlreadyAuthed: localStorage.get(`authed_${slugify(context.channelName, '_')}`) ? true : false
        });

        microsoftTeams.appInitialization.notifySuccess();
      });

      microsoftTeams.settings.registerOnSaveHandler((saveEvent) => {
        if(error) {
          saveEvent.notifyFailure(error.message)
        } else {
          saveEvent.notifySuccess()
        }
      })

      microsoftTeams.settings.registerOnRemoveHandler(async (removeEvent) => {
        try {
          const refresh_response = await service.getRefreshCode()
          const access_token = refresh_response.data.access_token
          await service.deregister(access_token)
          localStorage.clear()
          removeEvent.notifySuccess()
        } catch(e) {
          removeEvent.notifyFailure()
          this.setState({flowStep: false, error: e.message})
        }
      })
    }
  }

  render() {
    const {flowStep, error, needsGraphAuth, channelIsAlreadyAuthed} = this.state
    return (
      <Provider>
        {!channelIsAlreadyAuthed && !error && flowStep === 'pre-authentication' &&
        <div className="pre-auth">
          <div className="row-1">
            <div className="content">
              <div className="container">
                <div className="flex-container">
                  <div className="paragraph">
                    {error.error &&
                      <p className="error">Oops! Something went wrong. Please try again later!</p>
                    }
                    <p>Copy</p>
                    <OauthPopup
                      url={getAuthUrl()}
                      onCode={this.onCode}
                      onClose={this.onClose}
                    >
                      <button onClick={this.startFlow}></button>
                    </OauthPopup>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          }

        {channelIsAlreadyAuthed || (!error && flowStep === 'authenticated') && 
           <div className="success">
           <div className="row-1">
             <div className="paragraph">
               <div>Copy</div>
               <p>
                 <i> Copy</i>
               </p>
             </div>
           </div>

           <div className="row-2">
              <div className="row-background">
                 <div>Copy</div>
                  <ol>
                    <li>) Be sure to have authenticated with a <a href="https://teams.microsoft.com/">Microsoft Account</a></li>
                    <li>) Copy</li>
                  </ol>
              </div>
           </div>

           <div className="row-3">
              <img src={CompletionImage} alt="copy" />
              copy
           </div>
         </div>
        }


      </Provider>
    )
  }
}

export default App;
```

**Comment by aosolis** [comment]
Hello! `registerOnRemoveHandler` should be called from the [tab removal page](https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/create-tab-pages/removal-page), because that's the only page that will call that handler. That is why it's forbidden to call that method from the `settings` context.

**Comment by quesurifn** [comment]
@aosolis This is not a tab it's an Office 365 connector

**Comment by quesurifn** [comment]
registerOnRemove works in ComponentDidMount not componentWillMount

**Comment by skarahoda** [comment]
We have the same issue,

As far as I understand from code, `initialize` function sets frame context asynchronously. I noticed that the `microsoftTeams.settings.registerOnRemoveHandler` works when the initialize called and the context is not defined. That's why `microsoftTeams.settings.registerOnRemoveHandler` works for some cases.

**Comment by quesurifn** [comment]
@aosolis We added a removeUrl     
```
 microsoftTeams.settings.registerOnSaveHandler((saveEvent) => {
        const {channelId, channelName, teamId, hooplaToken, hooplaConnector, hooplaChannel} = this.state
        const contentUrl = `https://${window.location.href}?name={loginHint}&tenant={tid}&group={groupId}&theme={theme}`
        const entity = `hoopla-teams-${teamId}-${channelId}`;
        microsoftTeams.settings.setSettings({
            entityId: entity,
            contentUrl,
            configName: `Hoopla Integration with ${channelName}`,
            removeUrl: `${window.location.origin}?removal=true`
        });
```

however teams does not load the page. Why would this be?

**Comment by quesurifn** [comment]
Is there a requirement that the removalURL has to be different than the configUrl?

**Comment by quesurifn** [comment]
@Trinetra-MSFT

**Comment by Trinetra-MSFT** [comment]
@quesurifn , Let me check this, I will get back to you in sometime

**Comment by skarahoda** [comment]
@Trinetra-MSFT You can also check this: https://github.com/zeplin/microsoft-teams-app/blob/main/src/client/containers/ConfigurationUpdateContainer/hooks/useConfigurationDelete.ts. We patch `@microsoft/teams-js` to fix the issue [here](https://github.com/zeplin/microsoft-teams-app/blob/main/patches/%40microsoft%2Bteams-js%2B1.6.0.patch). I hope it will help for you.

**Comment by quesurifn** [comment]
@Trinetra-MSFT Any update on this?

**Comment by Trinetra-MSFT** [comment]
@quesurifn , you need to call the registerOnRemoveHandler after initialize. [microsoftTeams.settings.registerOnRemoveHandler().](https://github.com/OfficeDev/microsoft-teams-sample-connector-csharp/blob/master/TeamsToDoAppConnector/Views/Connector/Setup.cshtml) This event is getting triggered when user clicks on remove button in edit configuration page, 
![image](https://user-images.githubusercontent.com/50989436/97689107-1fd51280-1ac1-11eb-8d0f-fe0c57dd3655.png)
Here you can handle your db clean up but on click on final remove button it should remove the connector.
![image](https://user-images.githubusercontent.com/50989436/97689069-0cc24280-1ac1-11eb-95e0-59d19d592b26.png)

However we have work item in backlog to resolve this issue and allow remove in connector setting context

**Comment by quesurifn** [comment]
@Trinetra-MSFT 


```
 async componentWillMount() {
    const urlParams = new URLSearchParams(window.location.search);
    const inRemovalState = urlParams.get('removal')
    const inTokenState = urlParams.get('code')
    const insideTeams = await this.inTeams()

    if(insideTeams) {
      microsoftTeams.initialize()
      microsoftTeams.settings.registerOnRemoveHandler(async (removeEvent) => {
        try {
          await this.startCleanup()
          removeEvent.notifySuccess()
        } catch(e) {
          removeEvent.notifyFailure(e.message)
        }
      })
      microsoftTeams.getContext((context) => {
        const entity = `hoopla-teams-${context.teamId}-${context.channelId}`;
        this.setState({
          teamName: context.teamName,
          channelName: context.channelName,
          channelId: context.channelId,
          teamId: context.teamId,
          groupId: context.groupId,
          isAlreadyAuthed: localStorage.getItem(entity) ? true : false,
          flowStep: inRemovalState ? 'removal' : 'pre-authentication',
        });

        microsoftTeams.appInitialization.notifySuccess();
      });

      if(inTokenState) {
        await this.onCode(inTokenState)
      }

      microsoftTeams.settings.registerOnSaveHandler((saveEvent) => {
        const {channelId, channelName, teamId} = this.state
        const contentUrl = `https://${window.location.href}?name={loginHint}&tenant={tid}&group={groupId}&theme={theme}`
        const entity = `hoopla-teams-${teamId}-${channelId}`;
        microsoftTeams.settings.setSettings({
          entityId: entity,
          contentUrl,
          configName: `Hoopla Integration with ${channelName}`,
          removeURL: `${window.location.origin}?removal=true`
        });

        console.log('Remove URL')
        console.log(`${window.location.origin}?removal=true`)

        microsoftTeams.settings.getSettings(async (settings) => {
          try {
            const create_data = {...this.state, webhookUrl: settings.webhookUrl}
            await service.createChannel(create_data, access_token, iid)
            saveEvent.notifySuccess()
          } catch(e) {
            console.error("ERROR THROWN")
            console.log(e)
            saveEvent.notifyFailure(e.message)
          }
          });
       });
    }
  }

```

Yields 

```
Uncaught (in promise) Error: This call is not allowed in the 'settings' context
    at Object.t.ensureInitialized (MicrosoftTeams.min.js:1)
    at Object.e.registerOnRemoveHandler (MicrosoftTeams.min.js:1)
    at a.<anonymous> (App.js:91)
    at s (runtime.js:63)
    at Generator._invoke (runtime.js:293)
    at Generator.next (runtime.js:118)
    at n (asyncToGenerator.js:3)
    at c (asyncToGenerator.js:25)

```

In the console

**Comment by quesurifn** [comment]
subsequently wrapping the removeHandler in isRemovalState allows removes the console error but the removal modal doesn't load anything. 

Examples: 


```
if(insideTeams) {
      microsoftTeams.initialize()
      if(inRemovalState) {
        microsoftTeams.settings.registerOnRemoveHandler(async (removeEvent) => {
          try {
            await this.startCleanup()
            removeEvent.notifySuccess()
          } catch(e) {
            removeEvent.notifyFailure(e.message)
          }
        })
      } else {

        microsoftTeams.getContext((context) => {
          const entity = `teams-${context.teamId}-${context.channelId}`;
          this.setState({
            teamName: context.teamName,
            channelName: context.channelName,
            channelId: context.channelId,
            teamId: context.teamId,
            groupId: context.groupId,
            isAlreadyAuthed: localStorage.getItem(entity) ? true : false,
            flowStep: inRemovalState ? 'removal' : 'pre-authentication',
          });
  
          microsoftTeams.appInitialization.notifySuccess();
        });
  
        if(inTokenState) {
          await this.onCode(inTokenState)
        }
  
        microsoftTeams.settings.registerOnSaveHandler((saveEvent) => {
          const {channelId, channelName, teamId = this.state
          const contentUrl = `https://${window.location.href}?name={loginHint}&tenant={tid}&group={groupId}&theme={theme}`
          const entity = `teams-${teamId}-${channelId}`;
          microsoftTeams.settings.setSettings({
            entityId: entity,
            contentUrl,
            configName: `Integration with ${channelName}`,
            removeURL: `${window.location.origin}?removal=true`
          });
  
          console.log('Remove URL')
          console.log(`${window.location.origin}?removal=true`)
  
          microsoftTeams.settings.getSettings(async (settings) => {
            try {
              const create_data = {...this.state, webhookUrl: settings.webhookUrl}
              await service.createChannel(create_data, access_token, id)
              saveEvent.notifySuccess()
            } catch(e) {
              console.error("ERROR THROWN")
              console.log(e)
              saveEvent.notifyFailure(e.message)
            }
            });
         });
      }
    }
  }
```

**Comment by Trinetra-MSFT** [comment]
@quesurifn , 
I tried to create a fresh react app having connector following this [documentation](https://docs.microsoft.com/en-us/learn/modules/msteams-webhooks-connectors/7-exercise-o365-connectors), There is issue with the `registerOnRemoveHandler`  which is not allowed in setting context.  I will file a Bug for this and will track it.
```
 microsoftTeams.settings.registerOnRemoveHandler((removeEvent: microsoftTeams.settings.RemoveEvent) => {
              return true;
            });
```
![image](https://user-images.githubusercontent.com/50989436/97897618-bf620180-1d5c-11eb-94a7-f75b618bb742.png)

 As per documentation yomen generator sample code we are not calling `registerOnRemoveHandler `explicitly in app, while removing the connector it works fine however we put the removeHandler it gives error "This call is not allowed in setting context", but it removes the connector from channel configured list because it takes the same content url as remove URL & removes the connector from channel.

**Comment by quesurifn** [comment]
@Trinetra-MSFT Great. Do you have any idea how to get our React code to load the remove screen? Part of me worries if we submit the app it will be denied without this ability.

**Comment by Trinetra-MSFT** [comment]
Connector removal works fine once you configure connector in channel, Go to your configured connector list from channel, click on manage, You will be able to see the Remove screen for your connector on click of Remove button connector will be removing. That is already handled in sample code. You don't have write connector removal logic explicitly

**Comment by quesurifn** [comment]
@Trinetra-MSFT I don't believe this to be the case. The removeURL is supposed to show a removal page which we've done but doesn't display. The handleOnRemoveHandler is supposed to handle cleanup yet, the page doesn't load so the handler doesn't load. We will reference this conversation on app submission.

**Comment by Trinetra-MSFT** [comment]
@quesurifn , I've filed a Bug for registerOnRemoveHandler is not allowed in setting context. Will be updating you once it is fixed

**Comment by quesurifn** [comment]
Quick update, I have a clue as to why it's not loading. There's a cookies tab with `https://https` in dev tools which makes me think that something is up with the remove url.

<img width="264" alt="Screen Shot 2020-11-06 at 9 53 25 AM" src="https://user-images.githubusercontent.com/21092055/98380052-10e8e580-2016-11eb-97f9-8df235657728.png">

**Comment by Trinetra-MSFT** [comment]
@quesurifn , Can you please reach out to `microsoftteamsdev@microsoft.com` for your query so we can check the issue over call? I've tested by putting the removeURL same as content URL and able to see the [set up page](https://github.com/OfficeDev/microsoft-teams-sample-connector-csharp/blob/master/TeamsToDoAppConnector/Views/Connector/Setup.cshtml)

**Comment by skarahoda** [comment]
@Trinetra-MSFT there is a problem that when `microsoftTeams.settings.registerOnRemoveHandler` is called right after `microsoftTeams.initialize` it works since the context is not defined yet. However the function is called in callback of `microsoftTeams.initialize` then it fails.


You can try following code [in your example](https://github.com/OfficeDev/microsoft-teams-sample-connector-csharp/blob/master/TeamsToDoAppConnector/Views/Connector/Setup.cshtml) to see the bug.
```js
microsoftTeams.initialize(() => {
    microsoftTeams.settings.registerOnRemoveHandler(function (removeEvent) {
            var removeCalled = true;
        });
});
```

**Comment by quesurifn** [comment]
@Trinetra-MSFT I sent an email but haven't heard back

**Comment by Trinetra-MSFT** [comment]
@quesurifn , Didn't saw any email yet in my inbox , `microsoftteamsdev@microsoft.com` did you sent here only?

**Comment by Trinetra-MSFT** [comment]
@quesurifn , Hi Kyle, it will be good if you can respond to our mail and tell us your suitable time for the call.

**Comment by skarahoda** [comment]
> @Trinetra-MSFT there is a problem that when `microsoftTeams.settings.registerOnRemoveHandler` is called right after `microsoftTeams.initialize` it works since the context is not defined yet. However the function is called in callback of `microsoftTeams.initialize` then it fails.
> 
> You can try following code [in your example](https://github.com/OfficeDev/microsoft-teams-sample-connector-csharp/blob/master/TeamsToDoAppConnector/Views/Connector/Setup.cshtml) to see the bug.
> 
> ```js
> microsoftTeams.initialize(function () {
>     microsoftTeams.settings.registerOnRemoveHandler(function (removeEvent) {
>             var removeCalled = true;
>         });
> });
> ```

Hi @Trinetra-MSFT, can you reproduce the bug?

**Comment by Trinetra-MSFT** [comment]
SDK needs to be initialized before registering the remove handler else it will throw error.

**Comment by skarahoda** [comment]
Even when sdk is initialized, registiering remove handler produce `registerOnRemoveHandle is not allowed in the settings context` error. You can check the code below:

```js
microsoftTeams.initialize(function () {
    microsoftTeams.settings.registerOnRemoveHandler(function (removeEvent) {
            var removeCalled = true;
        });
});
```

At first initialization function is called. When initialization is done, it calls remove handler which is in the callback.

**Comment by piyumi25** [comment]
Hi,

I am also having same issue and I generated the connector app using latest version of yeoman generator. Looking for a solution for the issue.

**Comment by Trinetra-MSFT** [comment]
@piyumi25 : You can drop a email over `microsoftteamsdev@microsoft.com` so we can look into the issue

**Comment by gerrosc** [comment]
Hi! 

Any news on a fix for this issue?  I'm also getting the same problem.

Thanks!


By the way, I have a quick bit here that might help (or not) in finding what's causing this issue.  Some thing that i noticed;

When using version 1.5.0 of the script and do something like this pattern below, the error is not showing and everything seems fine;

```
           /// This works and no issue 

           microsoftTeams.initialize();
           microsoftTeams.settings.registerOnRemoveHandler((removeEvent: any) => {
                    ...
           });
           microsoftTeams.settings.registerOnSaveHandler((saveEvent: any) => {
                    ...
           });
```

However, when you introduce call in getting teams context like ```microsoftTeams.getContext(...)``` or something like this ```const [{ theme, context }] = useTeams();``` (from another component),  the code fails and error shows up. :-(

```
           ///  Examples
           /// - This has issue (or causes) -

          const [{ theme, context }] = useTeams();   // the useTeams()

          ///  - or something like this -

         microsoftTeams.initialize();
         microsoftTeams.getContext(function (context) {
                microsoftTeams.settings.registerOnRemoveHandler((removeEvent: any) => {
                        ...
                });
               microsoftTeams.settings.registerOnSaveHandler((saveEvent: any) => {
                        ...
               });
         });
```

On the latest version of script (i.e. 1.8.0). Even if you don't have ```microsoftTeams.getContext(...)``` call,  issue always occurs and page fails when it finds ```microsoftTeams.settings.registerOnRemoveHandler(...)``` code.

Hope that helps.

**Comment by piyumi25** [comment]
Hi,

Any update on this issue? I am getting stuck with the connector app because of the issue and waiting for a fix ASP.

**Comment by ydogandjiev** [comment]
I recently check in a fix for this issue and it was just released in the newly minted v1.9.0 version of the Client SDK:
https://www.npmjs.com/package/@microsoft/teams-js/v/1.9.0

Give it a shot!

**Comment by piyumi25** [comment]
> I recently check in a fix for this issue and it was just released in the newly minted v1.9.0 version of the Client SDK:
> https://www.npmjs.com/package/@microsoft/teams-js/v/1.9.0
> 
> Give it a shot!

Thanks the fix is working as expected.
